examples/tv-app/tv-common/clusters/messages/MessagesManager.h

```diff
@@ -21,6 +21,115 @@
+    constexpr static size_t kLabelMaxLength = 32;
+
+    CachedMessageOption(uint32_t id, std::string label) :
+        mLabel(label), mOption{ chip::MakeOptional(id), chip::MakeOptional(chip::CharSpan::fromCharString(mLabel.c_str())) }
+    {}
+
+    CachedMessageOption(const CachedMessageOption & option) :
+        mLabel(option.mLabel),
+        mOption{ option.mOption.messageResponseID, chip::MakeOptional(chip::CharSpan::fromCharString(mLabel.c_str())) } {};
+
+    CachedMessageOption & operator=(const CachedMessageOption & option) { return *this; };
+
+    chip::app::Clusters::Messages::MessageResponseOption GetMessageOption() { return mOption; }
+
+    ~CachedMessageOption() {}
+
+protected:
+    std::string mLabel;
+    chip::app::Clusters::Messages::MessageResponseOption mOption;
+};
+
+struct CachedMessage
+{
+    constexpr static size_t kMessageIdLength = 16;
+    constexpr static size_t kMaxOptionCount  = 4;
+
+    chip::ByteSpan mMessageId;
+
+    CachedMessage(const CachedMessage & message) :
+        mPriority(message.mPriority), mMessageControl(message.mMessageControl), mStartTime(message.mStartTime),
+        mDuration(message.mDuration), mMessageText(message.mMessageText), mOptions(message.mOptions)
+    {
+        memcpy(mMessageIdBuffer, message.mMessageIdBuffer, kMessageIdLength);
+        mMessageId = chip::ByteSpan(mMessageIdBuffer, kMessageIdLength);
+
+        for (CachedMessageOption & entry : mOptions)
+        {
+            mResponseOptions.push_back(entry.GetMessageOption());
+        }
+    };
+
+    CachedMessage & operator=(const CachedMessage & message) { return *this; };
+
+    CachedMessage(const chip::ByteSpan & messageId, const chip::app::Clusters::Messages::MessagePriorityEnum & priority,
+                  const chip::BitMask<chip::app::Clusters::Messages::MessageControlBitmap> & messageControl,
+                  const chip::app::DataModel::Nullable<uint32_t> & startTime,
+                  const chip::app::DataModel::Nullable<uint16_t> & duration, std::string messageText) :
+        mPriority(priority),
+        mMessageControl(messageControl), mStartTime(startTime), mDuration(duration), mMessageText(messageText)
+    {
+        if (messageId.size() != kMessageIdLength)
+        {
+            ChipLogProgress(Zcl, "CachedMessage bad message id size, ignoring");
+            return;
+        }
+        memcpy(mMessageIdBuffer, messageId.data(), kMessageIdLength);
+        mMessageId = chip::ByteSpan(mMessageIdBuffer, kMessageIdLength);
+    }
+
+    void AddOption(CachedMessageOption option)
+    {
+        mOptions.push_back(option);
+        mResponseOptions.push_back(option.GetMessageOption());
+    }
+
+    chip::app::Clusters::Messages::Structs::MessageStruct::Type GetMessage()
+    {
+        if (mResponseOptions.size() > 0)
+        {
+            chip::app::DataModel::List<chip::app::Clusters::Messages::MessageResponseOption> options(mResponseOptions.data(),
```

Comment:
Just noticed: why does the name of MessageResponseOption not match the spec?  That needs to get fixed ASAP, before we ship anything here.  Separate PR for this, obviously.

Reply:
I don't follow, it appears to me to follow the spec. See https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/app_clusters/Messages.adoc#ref_MessageResponseOptionStruct

Reply:
`MessageResponseOptionStruct` is not the same name as `MessageResponseOption`. Names have to match exactly.

Reply:
Yes.  Like I said: the name does not match.

Reply:
this comes from messages-delegate.h:

using MessageResponseOption = chip::app::Clusters::Messages::Structs::MessageResponseOptionStruct::Type;

But I can change that name to MessageResponseOptionStruct

Reply:
removed the using and replaced with complete path